File: java/day2_part2_2024.java

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Arrays;
import java.util.Scanner;

public class SafeReport {

    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(new File("input.txt"))) {
            int safeReportCount = 0;
            while (scanner.hasNextLine()) {
                int[] levels = parseLevels(scanner.nextLine());
                if (isSafeReport(levels) || isSafeWithOneRemoval(levels)) {
                    safeReportCount++;
                }
            }
            System.out.println(safeReportCount);
        } catch (FileNotFoundException e) {
            System.err.println("Error opening file: " + e.getMessage());
        }
    }

    static int[] parseLevels(String line) {
        return Arrays.stream(line.split("\\s+"))
                .mapToInt(Integer::parseInt)
                .toArray();
    }

    static boolean isSafeReport(int[] levels) {
        if (levels.length < 2) return false;
        int firstDiff = levels[1] - levels[0];
        if (firstDiff == 0) return false;
        boolean increasing = firstDiff > 0;
        for (int i = 0; i < levels.length - 1; i++) {
            int diff = levels[i + 1] - levels[i];
            if (diff == 0 || (increasing && diff <= 0) || (!increasing && diff >= 0) || Math.abs(diff) < 1 || Math.abs(diff) > 3)
                return false;
        }
        return true;
    }

    static boolean isSafeWithOneRemoval(int[] levels) {
        for (int i = 0; i < levels.length; i++) {
            int[] modifiedLevels = new int[levels.length - 1];
            System.arraycopy(levels, 0, modifiedLevels, 0, i);
            System.arraycopy(levels, i + 1, modifiedLevels, i, levels.length - i - 1);
            if (isSafeReport(modifiedLevels)) return true;
        }
        return false;
    }
}

File: elixir/day2_part2_2024.exs

defmodule Solver do
  def solve(filename) do
    filename
    |> File.stream!()
    |> Stream.map(&parse_levels/1)
    |> Stream.filter(&(&1 != nil))
    |> Stream.filter(&(is_safe_report?(&1) || is_safe_with_one_removal?(&1)))
    |> Enum.count()
    |> IO.puts()
  end

  defp parse_levels(line) do
    line
    |> String.trim()
    |> String.split()
    |> Enum.map(&String.to_integer/1)
    |> case do
      levels when is_list(levels) -> levels
      _ -> nil
    end
  end

  defp is_safe_report?(levels) do
    case levels do
      [_] -> false
      [a, b | _] ->
        diff = b - a
        if diff == 0 do
          false
        else
          (diff > 0 && is_increasing?(levels, abs(diff))) ||
          (diff < 0 && is_decreasing?(levels, abs(diff)))
        end

      _ ->
        false
    end
  end

  defp is_increasing?([_], _), do: true
  defp is_increasing?([a, b | rest], prev_diff) do
    diff = b - a
    abs_diff = abs(diff)
    if diff > 0 and abs_diff >= 1 and abs_diff <= 3 do
      is_increasing?([b | rest], abs_diff)
    else
      false
    end
  end

  defp is_decreasing?([_], _), do: true
  defp is_decreasing?([a, b | rest], prev_diff) do
    diff = b - a
    abs_diff = abs(diff)
    if diff < 0 and abs_diff >= 1 and abs_diff <= 3 do
      is_decreasing?([b | rest], abs_diff)
    else
      false
    end
  end

  defp is_safe_with_one_removal?(levels) do
    0..(length(levels) - 1)
    |> Enum.any?(fn i ->
      levels
      |> List.delete_at(i)
      |> is_safe_report?()
    end)
  end
end

Solver.solve("input.txt")
